### ISSUE TITLE
Witness passing quotation regression when using old libraries

### ISSUE THREAD

**Issue by krauthaufen** [dotnet/fsharp#11894]
When using SRTP-functions from old (pre witness-passing) libraries the runtime tries to resolve the `$W`-variant of the function and fails with something like

```
 System.InvalidOperationException: Could not bind function clamp$W in type Aardvark.Base.FSharpMath
 ```

**Repro steps**

1. Create a `net5.0` project
2. reference (for example) `Aardvark.Base.FSharp`
3. create a quotation using e.g. `clamp 0.0 1.0 x`  

Here's the full repro:
```fsharp
open Aardvark.Base

[<EntryPoint>]
let main argv =
    // fails here with
    // `System.InvalidOperationException: Could not bind function clamp$W in type Aardvark.Base.FSharpMath`
    let test = <@ fun (x : float) -> clamp 0.0 1.0 x @>
    printfn "%A" test
    0
```

**Expected behavior**

Should not fail and create old-style quotations (without the witnesses)

**Actual behavior**

Fails at runtime

**Known workarounds**

1. recompile all libraries with newer compiler 🙄
2. use older compiler/LangVersion for application

**Comment by dsyme** [comment]
Hmmm yes we will assume a library recompile here.

I expect we'll have to treat this as a point-in-time thing, given the two workarounds are available.

**Comment by krauthaufen** [comment]
Hey, yes I can recompile the particular library in question, nonetheless I think that this should be fixed in the compiler/FSharp.Core implementation, since it just breaks compatibility with older libraries.

I was under the impression that the general guideline for F# is something along the lines of "keep everything compatible when humanly possible" so I think this should at least be a warning and not just cause runtime exceptions.

I'd also be glad to help fixing that by creating regular `Call` expressions whenever the witness-methods are not available.
But that of course requires people to approve.

Cheers

**Comment by dsyme** [comment]
Yes, it's an inappropriate breaking change and not ideal, and in theory the logical thing to do is fix ti.

However a fix will take quite a while to get through the works, and becomes quite corner-case once "you have to be using quotations and a non-FSharp.Core library that contains inlined methods" is taken into account - most libraries of that character end up getting recompiled. And I guess it sort of simplifies our life from many perspectives if code is updated.

I'm also not entirely sure what the logic of the fix would be, I guess we would go looking in the .NET IL assembly to see if the `$W` methods are there.

**Comment by krauthaufen** [comment]
Sadly our packages are built by various people, so now I'll have to tell all of them to use at least some (yet to be found) compiler version and they won't be able to tell the difference until they use quotations with their binaries. I could of course validate that in a build-script (via searching for a particular `$W` method after the build) but all this causes lots of work which I would rather invest in fixing the issue.

Basically wr could simply catch the exception and create a `Call` expression like the old code did.

I'm aware that in a year or so this issue will likely not be around anymore and that I'll have to wait for several months to see the fix in production. However when someone sees this in 3 years no-one will that easily guess what caused it.

Cheers

**Comment by dsyme** [comment]
> some (yet to be found) compiler version

Just wondering about the "yet to be found" - you can tell them all to move to .NET SDK latest, that is anything supporting F# 5.0?

**Comment by krauthaufen** [comment]
Not really, the libs are netstandard, s.t. they can be used from old .netframework projects and can therefore be compile with older compilers

**Comment by krauthaufen** [comment]
By "yet to be found" I mean I'll have to find the F# compiler version that started emitting these methods, the dotnet version including said compiler and the Visual Studio Version including this dotnetn version

**Comment by dsyme** [comment]
Right, anything supporting F# 5.0 will do. I can see there is a problem if

1. Someone is publishing a component with an older toolchain
2. Someone is consuming a component using quotations etc. with a new toolchain assuming F# 5.0, adn they want to take advantage of F# 5.0 features

Anyway, I still think encouraging the publisher to update to a newer toolchain and republish seems best.  It seems best for it to happen sometime (partly because the witness passing may fix really problems in some quotation evaluation).

**Comment by krauthaufen** [comment]
Fair enough,

Btw are there any docs where I can find which VS/dotnet version includes F# >= 5.0?

VS2019 and dotnet 5 I assume?

**Comment by dsyme** [comment]
.NET SDK 5.0+

Visual Studio 16.7.0+ I believe (not certain).

**Comment by krauthaufen** [comment]
I know that there were blog entries all around but I'd have to binary search these whereas someone around here will just know that...

**Comment by dsyme** [comment]
This is resolved by recompiling the relevant packages as mentioned above
